### DISCUSSION TITLE
RuntimeError: Trying to backward through the graph a second time.

### DISCUSSION THREAD

**Discussion by isabel-s-rosa** [mir-group/nequip#170] [Q&A] [answered]
Hello!

I am calling a deployed nequip model from C++ using libtorch, but after I run one step of the physics engine and call the model again on the updated input, I see an error that looks like this:

```
Traceback of TorchScript, original code (most recent call last):
  File "/Users/isabelrosa/Library/Python/3.8/lib/python/site-packages/nequip/nn/_rescale.py", line 124, in forward
    def forward(self, data: AtomicDataDict.Type) -> AtomicDataDict.Type:
        data = self.model(data)
               ~~~~~~~~~~ <--- HERE
        if self.training:
            return data
  File "/Users/isabelrosa/Library/Python/3.8/lib/python/site-packages/nequip/nn/_grad_output.py", line 77, in forward
        data = self.func(data)
        # Get grads
        grads = torch.autograd.grad(
                ~~~~~~~~~~~~~~~~~~~ <--- HERE
            # TODO:
            # This makes sense for scalar batch-level or batch-wise outputs, specifically because d(sum(batches))/d wrt = sum(d batch / d wrt) = d my_batch / d wrt
RuntimeError: Trying to backward through the graph a second time (or directly access saved tensors after they have already been freed). Saved intermediate values of the graph are freed when you call .backward() or autograd.grad(). Specify retain_graph=True if you need to backward through the graph a second time or if you need to access saved tensors after calling backward.
```

Has anyone seen this before or have any idea how to go about fixing it? All I'm doing in my code is calling the `forward` function on the same model twice with different inputs. Thanks!

**Comment by Linux-cpp-lisp** [comment, marked as answer]
HI @isabel-s-rosa;

Hm... my first guess:

did you update your input tensor instead of making a new tensor?

**Reply by isabel-s-rosa** [reply]
The input to my model is a dictionary mapping strings to tensors, and both the dictionary and all the tensors are new. The only key-value pair I don't generate myself are the forces, which I get as output from the first calling of the model. Would that cause an issue?

**Reply by Linux-cpp-lisp** [reply]
If I understand you right, you are saying that you are inputing the forces from the previous call into the next call? Is this some kind of RNN setup?

**Reply by isabel-s-rosa** [reply]
Apologies, I misunderstood the model. I assumed forces were a required input but I realize now that I see the same behavior without passing in forces. To summarize the current problem, I am no longer inputting forces for either the first or second call, so everything being passed into the second call is totally new, and I still see the same error. Let me know what you think!

**Reply by Linux-cpp-lisp** [reply]
Aha, right--- the forces are only an output of the typical NequIP model.

Given this error, without seeing code, my best guess is still that the second call is getting something that is in fact not new--- the tensors themselves need to be new, not just their contents or the dict they are in.

**Reply by isabel-s-rosa** [reply]
My code for running the model is [here](https://github.com/isabel-s-rosa/meng/blob/main/npt-mtk.cpp), it's **very** messy still, but if you get a chance to look it over and let me know if you see any clear issues, that would be great!

**Reply by Linux-cpp-lisp** [reply]
Without looking through more closely can already ask: is any of the input to the second step connected in the autograd graph to the outputs of the first? For example, are the new positions a _PyTorch_ operation of old positions and predicted forces?

You can check for this kind of problem by checking the `requires_grad` in `input2` for all the tensors--- it shouldn't be true for any, but some might be true if they accidentally got connected in the autograd graph. To avoid this, use `.detach()`.

**Reply by isabel-s-rosa** [reply]
That's a good point! I just checked `requires_grad` on all the tensors that go into the second round of inputs, and it seems like everything is okay. Any other ideas? Thanks for all your help!

**Reply by isabel-s-rosa** [reply]
Never mind! This was my fault, I was passing in the wrong dictionary, thank you for all your help:)

**Reply by Linux-cpp-lisp** [reply]
Ah ok, glad to hear you found the issue!
